### PR TITLE
Allow building with `hashable-1.4.*`

### PR DESCRIPTION
**Pull request by RyanGlScott**
This was prompted by GaloisInc/parameterized-utils#126, but the changes needed on `what4`'s end can be applied independently.

### CODE REVIEW

File: what4/src/What4/Expr/App.hs [hunk 1]
```diff
@@ -937,7 +937,7 @@ instance TestEquality e => TestEquality (NonceApp t e) where
            ]
           )
 
-instance HashableF e => HashableF (NonceApp t e) where
+instance (HashableF e, TestEquality e) => HashableF (NonceApp t e) where
```
Comment (robdockins) [inline what4/src/What4/Expr/App.hs:940]: Why is this change necessary?

**Review by robdockins — APPROVED** [review]
Minor question about one change, but otherwise LGTM.

File: what4/src/What4/Expr/App.hs (hunk 1 above)
Comment (RyanGlScott) [reply what4/src/What4/Expr/App.hs:940]: The body of this `HashableF` invokes `hashWithSalt` on fields of type `Assignment` (in the `MapOverArrays` and `FnApp` cases). In GaloisInc/parameterized-utils#127, I had to add a `TestEquality` constraint to the `Hashable` instance for `Assignment`, so that constraint must be propagated through to this `HashableF` instance as a result.
